docs/references/expo/expo-impersonation.mdx

```diff
@@ -0,0 +1,509 @@
+            // this is the user ID of the use you're going to impersonate
+            user_id: userId,
+            actor: {
+              // this is the ID of the impersonator
+              sub: actorId,
+            },
+          }),
+        });
+
+        const data = await res.json();
+
+        setActor(data);
+      }
+    }
+
+    generateAndSetToken();
+  }, []);
+
+  return actor;
+}
+```
+
+### Create an API route to generate actor tokens
+
+Now create an endpoint that will call Clerk's [create actor token](/docs/reference/backend-api/tag/Actor-Tokens#operation/CreateActorToken) endpoint at `/actor_tokens` and pass in the Clerk secret key for authorization.
+
+```tsx {{ filename: "app/generateActorToken+api.tsx" }}
+export async function POST(request: Request) {
+  const body: { user_id: string; actor: { sub: string } } =
+    await request.json();
+
+  const { user_id, actor } = body;
+
+  const res = await fetch("https://api.clerk.com/v1/actor_tokens", {
+    method: "POST",
+    headers: {
+      Authorization: `Bearer ${process.env.CLERK_SECRET_KEY}`,
+      "Content-type": "application/json",
+    },
+    body: JSON.stringify({
+      user_id,
+      actor: {
+        sub: actor.sub,
+      },
+    }),
+  });
+
+  const data = await res.json();
+
+  return Response.json(data);
+}
+```
+
+### Create a hook to get impersonated user data
+
+Create a second custom hook called `useImpersonatedUser()` that will fetch data about the impersonator. You can use this to display UI only the impersonator will see for specific actions, like showing who is currently impersonating and matching the session to the right user for logout.
+
+```tsx {{ filename: "app/(dashboard)/index.tsx" }}
+function useImpersonatedUser(
+  actorSub: string,
+  setImpersonator: React.Dispatch<React.SetStateAction<string | UserDataJSON>>
+) {
+  React.useEffect(() => {
+    const getImpersonatedUser = async () => {
+      const res = await fetch(`/getImpersonatedUser`, {
+        method: "POST",
+        body: JSON.stringify({
+          impersonator_id: actorSub,
+        }),
+      });
+
+      const data = await res.json();
+
+      setImpersonator(data);
+
+      getImpersonatedUser();
+    };
+  }, [actorSub]);
+}
+```
+
+### Create an API route to retrieve user data
+
+With your hook setup you can now create an API route that will call Clerk's [retrieve user endpoint](https://clerk.com/docs/reference/backend-api/tag/Users#operation/GetUserList) at `/users` and get back the Impersonated user's full [`User` object](/docs/references/javascript/user/user#user-object).
+
+```tsx {{ filename: "app/getImpersonatedUser+api.tsx" }}
+export async function POST(request: Request) {
+  const body: { impersonator_id: string } = await request.json();
+
+  const { impersonator_id } = body;
+  const res = await fetch(`https://api.clerk.com/v1/users/${impersonator_id}`, {
+    headers: {
+      Authorization: `Bearer ${process.env.CLERK_SECRET_KEY}`,
```

Comment:
Is Expo safe with this? I don't think so -- I do not think we should be sending the secret key to the client even if it is Expo. This looks like a very dangerous block of code.

Reply:
This is inside an Expo API Route -- not the client code.

Reply:
Oh, this is the new feature. This should come with a callout then. 1, its still experimental. 2, it requires a host that supports it.

The callout should indicate something like 'For simplifying the code example, this is using the new, experimental Expo API Routes. Please review https://docs.expo.dev/router/reference/api-routes/ for more information. You can use your own API server instead of these Expo API routes.

Reply:
It's mentioned right in the `Before you start section 👍

Reply:
> 1. Because this example uses [`expo-router`](https://docs.expo.dev/router/installation/)'s experimental API routes, you must follow the [Expo docs](https://docs.expo.dev/router/reference/api-routes/#deployment) to deploy your routes to a domain. Doing so allows Clerk to access to your `CLERK_SECRET_KEY`, which you must pass to your impersonation endpoint.

Reply:
Using your own API server actualy _isn't_ that easy from what I had researched.

Reply:
Right, sorry. I should have written something more like the following. Feel free to take or leave the idea/thoughts as you see fit.

---

For simplifying the code example and getting started with development, this is using the new, experimental Expo API Routes. Please review https://docs.expo.dev/router/reference/api-routes/ for more information. You can use your own API server instead of these Expo API routes.

Both the Expo API Routes or using your own API server will introduce complexity to your application. You will need to make the choice on the ultimate implementation you go with.

Reply:
Thank you for the copy!

@alexisintech have any thoughts on moving this to a callout and maybe including some of the copy Roy wrote here? They basically only have the choice of using their experimental routes, unless they are ready to serve from a native running server, so I want to make sure they are aware that this would rely on some more complication no matter what.